### PR TITLE
[Exec] Split execution state interface - part 1

### PR DESCRIPTION
**Pull request by ramtinms**
This PR tries to split the execution state interface into multiple ones that could be later on satisfied by different implementations. 
Still, a lot to be done, I tried to split without that much change of logic for this PR. and I mostly concentrated on the register storage, the attestation storage interface requires way more major changes. i.e. moving the knowledge of state commitment and block header mapping to storage and not leaking it to the ingestion engine.

### CODE REVIEW

File: model/flow/ledger.go [hunk 1]
```diff
@@ -181,6 +181,18 @@ func (d RegisterEntries) Values() []RegisterValue {
 	return r
 }
 
+// RegisterIterator is used for iterating over a set of registers
+type RegisterIterator interface {
+	// First returns the first item of the iterator
+	First() *RegisterEntry
```
Comment (pattyshack) [inline model/flow/ledger.go:187]: rm.  HasNext should always be check first since the iterator yield an empty list

File: model/flow/ledger.go [hunk 2]
```diff
@@ -181,6 +181,18 @@ func (d RegisterEntries) Values() []RegisterValue {
...
+	// HasNext returns true if another register could be read
+	HasNext() bool
+	// Next returns the next register and moves the cursor
+	Next() *RegisterEntry
```
Comment (zhangchiqing) [inline model/flow/ledger.go:191]: can we combine the `First` and `Next`, as well as `HasNext` into a single:
```
Next() (firstOrNext *RegisterEntry, hasNext bool)
```

File: engine/execution/state/state.go [hunk 3]
```diff
@@ -19,10 +18,65 @@ import (
 	"github.com/onflow/flow-go/storage/badger/procedure"
 )
 
-// ReadOnlyExecutionState allows to read the execution state
-type ReadOnlyExecutionState interface {
-	// NewStorageSnapshot creates a new ready-only view at the given state commitment.
-	NewStorageSnapshot(flow.StateCommitment) fvmState.StorageSnapshot
+// SnapshotReader provides read functionality at a specific snapshot
```
Comment (zhangchiqing) [inline engine/execution/state/state.go:21]:
```suggestion
// SnapshotReader provides read functionality at a specific snapshot of the execution state
```

File: engine/execution/state/state.go [hunk 4]
```diff
@@ -19,10 +18,65 @@ import (
...
+	// Implementation must be concurrency safe;
+	// and must handle repetition of the same events (with some processing overhead).
+	OnBlockExecuted(block *flow.Header, result *execution.ComputationResult) error
+}
+
+type ReadOnlyRegisterStorage interface {
+	module.ReadyDoneAware
+	// SnapshotReaderAtBlock returns an SnapshotReader for the given blockID
+	// if a register is not changed in the given block the latest value from ancestor blocks is returned
+	// it could be used for reading several values of the same blockID
```
Comment (zhangchiqing) [inline engine/execution/state/state.go:65]: need to define error for cases where blockID is not found, or not executed.

File: engine/execution/state/state.go [hunk 5]
```diff
@@ -19,10 +18,65 @@ import (
...
+	SnapshotReaderAtBlock(blockID flow.Identifier) (SnapshotReader, error)
+
+	// TODO(ramtin): consolidate this into above and depricate
+	NewStorageSnapshot(commitment flow.StateCommitment) SnapshotReader
+}
+
+// RegisterStorage stores registers at given height,
```
Comment (zhangchiqing) [inline engine/execution/state/state.go:72]: | at given height

Should be "at given block", right? we need to be fork-aware

**Pull request closed without merging**